fix: revive hangs on Windows if go.mod is not found

alexandear opened the pull request:
The PR fixes #1120.

The problem is the `retrieveModFile` function, which hangs forever when `go.mod` is not found. `filepath.Dir` returns `C:/` on Windows, but the implementation expects `"."` or `"/"`.

The bug appeared after changing `path.Dir` to `filepath.Dir` in #1073. We can revert to using `path.Dir`, but a better solution is to add `strings.TrimPrefix(dir, filepath.VolumeName(dir))` for Windows.

chavacava commented:
Hi @alexandear, thanks for investigating the problem and proposing a fix.

About: _We can revert to using `path.Dir`, but a better solution is to add `strings.TrimPrefix(dir, filepath.VolumeName(dir))` for Windows._

Why do you prefer adding OS-dependent code instead of reverting to `path.Dir`?

alexandear commented:
> Hi @alexandear, thanks for investigating the problem and proposing a fix.
> 
> About: _We can revert to using `path.Dir`, but a better solution is to add `strings.TrimPrefix(dir, filepath.VolumeName(dir))` for Windows._
> 
> Why do you prefer adding OS-dependent code instead of reverting to `path.Dir`?

1. `path/filepath` is preferable for working with cross-platform paths. This is by design of the standard package.
2. Changing to `filepath` fixed a bug with different behavior of `retrieveModFile` on Unix and Windows.

For example, add `fmt.Println("retrieveModFile dir: ", dir)` inside the `for` loop in `retrieveModFile`.

And run `revive` on Mac:

```sh
❯ ./revive/revive t.go
retrieveModFile dir:  /Users/Oleksandr_Redko/src/github.com/mgechev
retrieveModFile dir:  /Users/Oleksandr_Redko/src/github.com
retrieveModFile dir:  /Users/Oleksandr_Redko/src
retrieveModFile dir:  /Users/Oleksandr_Redko
retrieveModFile dir:  /Users
retrieveModFile dir:  /
t.go:1:1: invalid file t.go: t.go:1:1: expected 'package', found 'func'
```

Run `revive` on Mac:

```pwsh
PS C:\Users\Oleksandr_Redko\src\github.com\mgechev> revive.exe t.go
retrieveModFile dir:  C:\Users\Oleksandr_Redko\src\github.com\mgechev
retrieveModFile dir:  .
t.go:1:1: invalid file t.go: t.go:1:1: expected 'package', found 'func'
```

As we see, `retrieveModFile` searches for the `go.mod` on Mac inside 6 directories, and on Windows only inside 1 directory.


With `filepath.Dir` and the fix in this PR:

```pwsh
PS C:\Users\Oleksandr_Redko\src\github.com\mgechev> revive.exe t.go
retrieveModFile dir:  C:\Users\Oleksandr_Redko\src\github.com\mgechev
retrieveModFile dir:  \Users\Oleksandr_Redko\src\github.com
retrieveModFile dir:  \Users\Oleksandr_Redko\src
retrieveModFile dir:  \Users\Oleksandr_Redko
retrieveModFile dir:  \Users
retrieveModFile dir:  \
t.go:1:1: invalid file t.go: t.go:1:1: expected 'package', found 'func'
```

Actually, I refactored the fix and removed `if runtime.GOOS == "windows"`, please recheck.

alexandear commented:
GitHub Actions workflow for running tests on Windows added in #1125
